### ISSUE TITLE
Unclear error message when trying to authenticate with wrong password: E409 Conflict

### ISSUE THREAD

**Issue by sk1talets** [verdaccio/verdaccio#441]
#### My reason:

I'm getting the following error message from NPM 5 while trying to authenticate with a wrong password:
```
[mac ~]# npm login --registry http://localhost:4873
Username: username
Password: 
Email: (this IS public) user@email.com
npm ERR! code E409
npm ERR! Conflict

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/user/.npm/_logs/2017-12-20T12_39_01_443Z-debug.log
[mac ~]# 
```

#### Steps to reproduce:

Try to `npm login` with an incorrect password.

####  App Version:

Verdaccio 2.7.0
NodeJS v8.9.3
NPM 5.5.1

####  Config file:	

```
storage: /storage
uplinks:
  npmjs:
    url: https://registry.npmjs.org/
    auth:
      type: bearer
      token: xxx 
packages:
  '@*/*':
    access: $authenticated
    publish: $authenticated
    proxy: npmjs
  '**':
    access: $all
    publish: $authenticated
    proxy: npmjs
logs:
  - {type: stdout, format: pretty, level: info}
auth:
  ldap:
    type: ldap
    groupNameAttribute: cn
    client_options:
      url: ldaps://local.ad
      adminDn: admin@ad.net
      adminPassword: xxxx
      searchBase: OU=users,DC=ad,DC=net
      searchFilter: (sAMAccountName={{username}})
      searchAttributes:
        - memberOf
````

#### Additional information:

It seems that NPM 5 expects 401 status code and the `www-authenticate` header to be set to display a proper error message:
```
[mac ~]# npm login --registry https://registry.npmjs.org
Username: username
Password: 
Email: (this IS public) user@email.com
npm ERR! code E401
npm ERR! Incorrect or missing password.
npm ERR! If you were trying to login, change your password, create an
<...>
```
but Verdaccio overrides response status code to be 409.

#### Additional verbose log:

Verdaccio log:
```
<...>
 http <-- 409, user: undefined(127.0.0.1), req: 'PUT /-/user/org.couchdb.user:username', error: bad username/password, access denied
 info <-- 127.0.0.1 requested 'GET /-/user/org.couchdb.user:username?write=true'
 http <-- 200, user: undefined(127.0.0.1), req: 'GET /-/user/org.couchdb.user:username?write=true', bytes: 0/53
 info <-- 127.0.0.1 requested 'PUT /-/user/org.couchdb.user:username/-rev/undefined'
 warn --- LDAP error no such user: "username"
 warn --- LDAP error no such user: "username"
 http <-- 409, user: undefined(127.0.0.1), req: 'PUT /-/user/org.couchdb.user:username/-rev/undefined', error: bad username/password, access denied
 info <-- 127.0.0.1 requested 'PUT /-/user/org.couchdb.user:username'
 warn --- LDAP error no such user: "username"
 http <-- 409, user: undefined(127.0.0.1), req: 'PUT /-/user/org.couchdb.user:username', error: bad username/password, access denied
<...>
```

**Comment by juanpicado** [comment]
Please, fix the format.

**Comment by sk1talets** [comment]
Done. You are so quick 🥇

**Comment by juanpicado** [comment]
not always 🙈. Do not tell it too loud. I'll take a look later. @ayusharma keep in mind this for the external plugin, please.

**Comment by ayusharma** [comment]
Okay . I'll keep in mind @juanpicado :)

**Comment by sk1talets** [comment]
Btw, if an auth plugin implements only the `authenticate` method, I get `409 error: registration is disabled` from the default plugin (or whatever you call it) which is also a bit misleading
```
 info <-- 127.0.0.1 requested 'PUT /-/user/org.couchdb.user:username'
 http <-- 409, user: undefined(127.0.0.1), req: 'PUT /-/user/org.couchdb.user:username', error: registration is disabled
 info <-- 127.0.0.1 requested 'GET /-/user/org.couchdb.user:username?write=true'
 http <-- 200, user: undefined(127.0.0.1), req: 'GET /-/user/org.couchdb.user:username?write=true', bytes: 0/53
 info <-- 127.0.0.1 requested 'PUT /-/user/org.couchdb.user:username/-rev/undefined'
 warn --- LDAP error no such user: "username"
 http <-- 409, user: undefined(127.0.0.1), req: 'PUT /-/user/org.couchdb.user:username/-rev/undefined', error: registration is disabled
 info <-- 127.0.0.1 requested 'PUT /-/user/org.couchdb.user:username'
 http <-- 409, user: undefined(127.0.0.1), req: 'PUT /-/user/org.couchdb.user:username', error: registration is disabled
```

This is valid for `verdaccio-ldap` and `verdaccio-activedirectory` plugins.
Adding an empty `adduser` method helps:
```
[mac ~/Work/registry-proxy]# grep VerdaccioLdap start.js 
const VerdaccioLdap = require('verdaccio-ldap');
		VerdaccioLdap.prototype.adduser = (u, p, cb) => cb(null, true);
[mac ~/Work/registry-proxy]#
```

**Comment by juanpicado** [comment]
@sk1talets thanks for reporting, I checked and you are right. 

On load `auth` plugins we do not check whether `add_user` method exists 

```
this.plugins = loadPlugin(config, config.auth, plugin_params, function(p) {
      return p.authenticate || p.allow_access || p.allow_publish;
    });
```
And we register this plugin at the end of the chain.

```
 this.plugins.push({
      authenticate: function(user, password, cb) {
        return cb( Error[403]('bad username/password, access denied') );
      },

      add_user: function(user, password, cb) {
        return cb( Error[409]('registration is disabled') );
      },

      allow_access: allow_action('access'),
      allow_publish: allow_action('publish'),
    });
```
So according this logic 
```
if (typeof(p[n]) !== 'function') {
        next();
      } else {
        // p.add_user() execution
        p[n](user, password, function(err, ok) {
          if (err) return cb(err);
          if (ok) return self.authenticate(user, password, cb);
          next();
        });
      }
```
If the method `add_user` is not there, we skip the plugin and we go to the next one, so, the last will return `Error[409]('registration is disabled')`

To fix this, we need to update the `auth.js` logic and ask always for `add_user` method. This is a kinda breaking change since some plugins will stop to work if we fix it should be done only in `3.x`

**Comment by juanpicado** [comment]
@sk1talets and the issue with `401` handle but the plugin itself, that will be fixed here https://github.com/verdaccio/verdaccio-htpasswd/issues/5

**Comment by juanpicado** [comment]
Fix available on `verdaccio@3.0.0-alpha.11`
